spec/system/user_edits_an_article_spec.rb

```diff
@@ -0,0 +1,52 @@
+require "rails_helper"
+
+RSpec.describe "User edits an article" do
+  it "shows an edit button on the article show page" do
+    user = create(:user)
+    visit root_path(as: user)
+    fill_in "Title", with: "This is my awesome title"
+    fill_in "Content", with: "My awesome content"
+    click_button "Submit"
+    article = Article.last
+    visit article_path(as: user, id: article.id)
+
+    expect(page).to have_selector(:link_or_button, "Edit")
+  end
+
+  context "user clicks edit button" do
+    it "shows the edited article" do
+      user = create(:user)
+      visit root_path(as: user)
+      fill_in "Title", with: "This is my awesome title"
+      fill_in "Content", with: "My awesome content"
+      click_button "Submit"
+      article = Article.last
+      visit edit_article_path(as: user, id: article.id)
+      fill_in "Title", with: "This is my edited awesome title"
+      fill_in "Content", with: "My edited awesome content"
+      click_button "Submit"
+
+      expect(page).to have_content "Article was successfully edited"
+      edited_article = Article.last
+      expect(page).to have_content "Article ##{edited_article.id}"
+      expect(page).to have_content edited_article.title
+      expect(page).to have_content edited_article.content
+    end
```

Comment:
**suggestion: (blocking):** Rewrite this spec.

I would rewrite this spec as follows.

```ruby
  it "shows the edited article" do
    user = create :user
    article = create(
      :article,
      title: "This my awesome title",
      content: "My awesome content"
    )

    visit edit_article_path article, as: user
    fill_in "Title", with: "This is my edited awesome title"
    fill_in "Content", with: "My edited awesome content"
    click_button "Submit"

    expect(page).to have_content "Article was successfully edited"
    expect(page).to have_content "Article ##{article.id}"
    expect(page).to have_content "This is my edited awesome title"
    expect(page).to have_content "My edited awesome content"
  end
```

A lot is happening here, so I will go through it step-by-step.

```diff
-    visit root_path(as: user)
-    fill_in "Title", with: "This is my awesome title"
-    fill_in "Content", with: "My awesome content"
-    click_button "Submit"
-    article = Article.last
+    article = create(
+      :article,
+      title: "This my awesome title",
+      content: "My awesome content"
+    )
```

The first thing I did was remove the unnecessary visit to the "New Article" page. We can assume that our author has created an article and that it exists in the database. Interacting with the browser in our tests is slow, and we want to make our tests as fast as possible.

```diff
-    visit edit_article_path(as: user, id: article.id)
+    visit edit_article_path article, as: user
```

Next, I changed the call to `edit_article_path`. If you pass an object as the first argument to this method, Rails will figure out its ID. There is no need for it to be an Active Record object. It could also be any sort of object. Under the hood, Rails checks to see if the object responds to an `#id` method.

```diff
-    edited_article = Article.last
-    expect(page).to have_content "Article ##{edited_article.id}"
-    expect(page).to have_content edited_article.title
-    expect(page).to have_content edited_article.content
+    expect(page).to have_content "Article ##{article.id}"
+    expect(page).to have_content "This is my edited awesome title"
+    expect(page).to have_content "My edited awesome content"
```

The next step was to rewrite our assertions. We have to remember that our tests tell a story. We want them to be as descriptive as possible. We want to prevent readers from scanning around the code, looking for specific values. Let's make everything simple and put it right there. This also allows us to remove the unnecessary secondary call to the database.